Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -272,24 +311,26 @@ public sealed class Server : ServerBase, IDisposable {
                 // Pick up any changes on disk that we didn't know about
                 EnqueueItem(doc, AnalysisPriority.Low);
             }
-            return Task.CompletedTask;
         }
 
+
         public override async Task DidChangeConfiguration(DidChangeConfigurationParams @params) {
+            await _analyzerCreationTcs.Task;
             if (_analyzer == null) {
                 LogMessage(MessageType.Error, "change configuration notification sent to uninitialized server");
                 return;
             }
 
-            await _analyzer.ReloadModulesAsync();
+            _queue.Enqueue(new AnalysisQueueWorkItem(() => _analyzer.ReloadModulesAsync().WaitAndUnwrapExceptions()), AnalysisPriority.Normal);
```

Comment:
This looks dangerous.
- `_queue.Enqueue` won't execute `ReloadModulesAsync` immediately. If there is any code that awaits `DidChangeConfiguration` to ensure that modules are reloaded, it will be broken
- `AnalysisQueue` will reorder project entries, but it won't reorder `AnalysisQueueWorkItem` cause they all are considered different. This is good in the way that `ProjectEntry.Analyze` will be called always after `ReloadModulesAsync`, but has a side effect that several concurrent calls of `ReloadModulesAsync` will be executed one by one instead of swallowing all but first and last call.
- `ReloadModulesAsync` may start new tasks and await on them while `AnalysisQueue` is thread-bound. `AnalysisSynchronizationContext` doesn't support reentrancy and `AnalysisQueue` thread is blocked until method is completed, which means that if anywhere on the code path we have missed `ConfigureAwait(false)` or provide any other way for `AnalysisSynchronizationContext` to be captured, we'll get a deadlock.

Also, `ReloadModulesAsync` should handle `CancellationToken` by itself, but that is a separate issue.

Reply:
Well, I am open to suggestions. Currently it is broken b/c ReloadModulesAsync is called from two threads while
```
        /// This method should be called on the analysis thread and is usually invoked
        /// when the interpreter signals that it's modules have changed.
```
So basically change is to queue item before calling `_queue.Enqueue` in `DidChangeConfiguration`. Nothing awaits on DidChangeConfiguration in reality, it is out of proc call, we can make it void.

`ReloadModulesAsync` should never be called in parallel AFAIU

Reply:
I think analyzer should probably handle reload internally in its worker thread rather than caller know what is permitted and what not. This will be a bigger change.

Reply:
If it is queued, then `_reloadLock.WaitAsync()` prob not needed as there are no concurrent calls (apart of one at the creation at `CreateAsync`. I think this one can be queued too theoretically.

Now, we can also avoid doing anything in `DidChangeConfiguration`. Changing interpreter will actually restart the server.

Reply:
Ok, here is a version of small fix: https://github.com/AlexanderSher/PTVS/commit/bf068bdc4b0b6be3eb34cf3753996f705796e58c
- `DidChangeConfiguration` will wait for last `ReloadModulesAsync` to be processed
- `AnalysisQueue` will properly reorder `ReloadModulesAsync` call if it is already scheduled
- `ReloadModulesAsync` will be started in thread-pool so reentrancy shouldn't be a problem.